quantization/src/encoded_vectors_binary.rs

```diff
@@ -88,39 +134,89 @@ impl<TStorage: EncodedStorage> EncodedVectorsBin<TStorage> {
         #[cfg(all(target_arch = "aarch64", target_feature = "neon"))]
         if std::arch::is_aarch64_feature_detected!("neon") {
             unsafe {
-                return impl_xor_popcnt_neon(
-                    v1.as_ptr() as *const u64,
-                    v2.as_ptr() as *const u64,
+                return impl_xor_popcnt_neon_uint128(
+                    v1.as_ptr() as *const u8,
+                    v2.as_ptr() as *const u8,
                     v1.len() as u32,
                 ) as usize;
             }
         }
 
         let mut result = 0;
-        for (b1, b2) in v1.iter().zip(v2.iter()) {
+        for (&b1, &b2) in v1.iter().zip(v2.iter()) {
             result += (b1 ^ b2).count_ones() as usize;
         }
         result
     }
 
-    /// Estimates how many `StorageType` elements are needed to store `size` bits
     fn get_storage_size(size: usize) -> usize {
-        let mut result = size / BITS_STORE_TYPE_SIZE;
-        if size % BITS_STORE_TYPE_SIZE != 0 {
+        let bits_count = 8 * std::mem::size_of::<Self>();
+        let mut result = size / bits_count;
+        if size % bits_count != 0 {
             result += 1;
         }
         result
     }
+}
+
+impl<TBitsStoreType: BitsStoreType, TStorage: EncodedStorage>
+    EncodedVectorsBin<TBitsStoreType, TStorage>
+{
+    pub fn encode<'a>(
+        orig_data: impl Iterator<Item = impl AsRef<[f32]> + 'a> + Clone,
+        mut storage_builder: impl EncodedStorageBuilder<TStorage>,
+        vector_parameters: &VectorParameters,
+        stop_condition: impl Fn() -> bool,
+    ) -> Result<Self, EncodingError> {
+        debug_assert!(validate_vector_parameters(orig_data.clone(), vector_parameters).is_ok());
+
+        for vector in orig_data {
+            if stop_condition() {
+                return Err(EncodingError::Stopped);
+            }
+
+            let encoded_vector = Self::encode_vector(vector.as_ref());
+            let encoded_vector_slice = encoded_vector.encoded_vector.as_slice();
+            let bytes = transmute_to_u8_slice(encoded_vector_slice);
+            storage_builder.push_vector_data(bytes);
+        }
+
+        Ok(Self {
+            encoded_vectors: storage_builder.build(),
+            metadata: Metadata {
+                vector_parameters: vector_parameters.clone(),
+            },
+            bits_store_type: PhantomData,
+        })
+    }
+
+    fn encode_vector(vector: &[f32]) -> EncodedBinVector<TBitsStoreType> {
+        let mut encoded_vector =
+            vec![Default::default(); TBitsStoreType::get_storage_size(vector.len())];
```

Comment:
Just realised that for `u128` we may have the wrong alignment here. Need additional investigation but it's not a part of this PR